build-tools/packages/build-tools/src/common/fluidUtils.ts

```diff
@@ -31,35 +32,72 @@ async function isFluidRootPackage(dir: string) {
 	return false;
 }
 
-async function inferRoot() {
-	let fluidConfig = findUp.sync("fluidBuild.config.cjs", { cwd: process.cwd(), type: "file" });
-	if (fluidConfig === undefined) {
-		traceInit(`No fluidBuild.config.cjs found. Falling back to git root.`);
+async function inferRoot(buildRoot: boolean) {
+	const config = await findUp("fluidBuild.config.cjs", {
+		cwd: process.cwd(),
+		type: "file",
+	});
+	if (config !== undefined) {
+		return path.dirname(config);
+	}
+
+	traceInit(`No fluidBuild.config.cjs found. Falling back to git root.`);
+	try {
 		// Use the git root as a fallback for older branches where the fluidBuild config is still in
 		// package.json
 		const gitRoot = childProcess
-			.execSync("git rev-parse --show-toplevel", { encoding: "utf8" })
+			.execSync("git rev-parse --show-toplevel", {
+				encoding: "utf8",
+				stdio: ["ignore", "pipe", "ignore"],
+			})
 			.trim();
-		fluidConfig = path.join(gitRoot, "package.json");
-		if (fluidConfig === undefined || !existsSync(fluidConfig)) {
-			return undefined;
+
+		const gitRootPackageJson = path.join(gitRoot, "package.json");
+		if (existsSync(gitRootPackageJson)) {
+			if (!buildRoot) {
+				return gitRoot;
+			}
+			// For build root, we require it to have fluidBuild property.
+			const parsed = await readJson(gitRootPackageJson);
+			if (parsed.fluidBuild !== undefined) {
+				return gitRoot;
+			}
 		}
+	} catch (e) {
+		traceInit(`Error getting git root: ${e}`);
 	}
-	const isRoot = await isFluidRootPackage(path.dirname(fluidConfig));
-	if (isRoot) {
-		return path.dirname(fluidConfig);
+
+	if (buildRoot) {
+		// For fluid-build, just use the enclosing workspace or package if exists
+		try {
+			traceInit(`No git root found. Trying enclosing workspace/package`);
+			const { rootDir } = await getPackages(process.cwd());
+			return rootDir;
+		} catch (e) {
+			traceInit(`Error getting packages: ${e}`);
+		}
 	}
 
 	return undefined;
 }
 
-export async function getResolvedFluidRoot() {
+async function inferFluidRoot(buildRoot: boolean) {
```

Comment:
What's the difference between inferRoot and inferFluidRoot? Why are both needed?

Reply:
I refactor the code into two functions so that I can do early return pattern in `inferFluidRoot`, so that we can add a final check for all the possible returns from `inferRoot`